File: src/main/java/com/taw/polybank/service/CurrencyExchangeService.java
package com.taw.polybank.service;

import com.taw.polybank.dao.CurrencyExchangeRepository;
import com.taw.polybank.dto.CurrencyExchangeDTO;
import com.taw.polybank.entity.CurrencyExchangeEntity;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@Service
public class CurrencyExchangeService {

    @Autowired
    protected CurrencyExchangeRepository currencyExchangeRepository;

    public void save(CurrencyExchangeDTO currencyExchangeDTO, BadgeService badgeService) {
        CurrencyExchangeEntity currencyExchange = this.toEntity(currencyExchangeDTO, badgeService);
        currencyExchangeRepository.save(currencyExchange);
        currencyExchangeDTO.setId(currencyExchange.getId());

        // TODO add transaction to currency exchange
        // TODO add payment to currency exchange
    }

    public CurrencyExchangeEntity toEntity(CurrencyExchangeDTO currencyExchangeDTO, BadgeService badgeService) {
        CurrencyExchangeEntity currencyExchange = currencyExchangeRepository.findById(currencyExchangeDTO.getId()).orElse(new CurrencyExchangeEntity());
        currencyExchange.setId(currencyExchangeDTO.getId());
        currencyExchange.setInitialAmount(currencyExchangeDTO.getInitialAmount());
        currencyExchange.setFinalAmount(currencyExchangeDTO.getFinalAmount());
        currencyExchange.setBadgeByInitialBadgeId(badgeService.toEntity(currencyExchangeDTO.getBadgeByInitialBadgeId()));
        currencyExchange.setBadgeByFinalBadgeId(badgeService.toEntity(currencyExchangeDTO.getBadgeByFinalBadgeId()));
        return currencyExchange;
    }
}

File: src/main/java/com/taw/polybank/service/CurrencyExchangeService.kt
package com.taw.polybank.service

import com.taw.polybank.dao.CurrencyExchangeRepository
import com.taw.polybank.dto.CurrencyExchangeDTO
import com.taw.polybank.entity.CurrencyExchangeEntity
import org.springframework.stereotype.Service

/**
 * @author Illya Rozumovskyy
 */
@Service
class CurrencyExchangeService(private val currencyExchangeRepository: CurrencyExchangeRepository) {

  fun save(currencyExchangeDTO: CurrencyExchangeDTO, badgeService: BadgeService) {
    val currencyExchange = this.toEntity(currencyExchangeDTO, badgeService)
    currencyExchangeRepository.save<CurrencyExchangeEntity>(currencyExchange)
    currencyExchangeDTO.id = currencyExchange.id
  }

  fun toEntity(
    currencyExchangeDTO: CurrencyExchangeDTO,
    badgeService: BadgeService,
  ): CurrencyExchangeEntity {
    val currencyExchange =
      currencyExchangeRepository
        .findById(currencyExchangeDTO.getId())
        .orElse(CurrencyExchangeEntity())
    currencyExchange.setId(currencyExchangeDTO.getId())
    currencyExchange.setInitialAmount(currencyExchangeDTO.getInitialAmount())
    currencyExchange.setFinalAmount(currencyExchangeDTO.getFinalAmount())
    currencyExchange.setBadgeByInitialBadgeId(
      badgeService.toEntity(currencyExchangeDTO.getBadgeByInitialBadgeId()),
    )
    currencyExchange.setBadgeByFinalBadgeId(
      badgeService.toEntity(currencyExchangeDTO.getBadgeByFinalBadgeId()),
    )
    return currencyExchange
  }
}
